### PR TITLE
[API Compatibility] paddle.randint

### PR DESCRIPTION
**Pull request by Manfredss**
### PR Category
User Experience

### PR Types
Improvements

### Description
Add ``isinstance(high, (list, tuple))`` check in ``paddle.randint`` to support PyTorch-style two-positional-arg calling pattern ``randint(high, size)``, where the second arg is a ``shape`` instead of an ``int``. 

### 是否引起精度变化
否

### CODE REVIEW

File: python/paddle/tensor/random.py [hunk 1]
```diff
@@ -1861,7 +1861,11 @@ def randint(
             >>> # doctest: -SKIP
 
     """
-    if high is None:
+    if isinstance(high, (list, tuple)):
```
Comment (zhwesky2010) [inline python/paddle/tensor/random.py:1877]: 这个其实类似一个简单的装饰器重载效果。给这个randint加一个 `@overload`注解吧，提升API签名可读性，按Paddle签名、Pytorch签名来写，参考其他overload，例如reshape、index_add、var等

**Comment by algorithm1832** [comment]
`/python/paddle/nn/functional/loss.py`的Line 2418
`/python/paddle/nn/functional/common.py`的Line 2768

这两个文件内的示例代码使用到了`paddle.randint`，调用的`high`不符合overload，所以static失败了

可能需要一并改一下

File: python/paddle/tensor/random.py [hunk 2]
```diff
@@ -1765,6 +1765,29 @@ def uniform_(
     return _C_ops.uniform_inplace_(x, min, max, seed, 0, 0, 1.0)
 
 
+@overload
```
Comment (zhwesky2010) [inline python/paddle/tensor/random.py:1732]: https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference/torch_more_args/torch.randint.html

这个API的对齐可能没那么简单，缺的参数里，除了generator、layout、memory_format这三个外，其他都需要新增。

可以参考下 rand、randn等其他随机数API的修改。

Comment (Manfredss) [reply python/paddle/tensor/random.py:1732]: 好的，我看看

File: python/paddle/tensor/random.py [hunk 3]
```diff
@@ -1875,19 +1916,41 @@ def randint(
...
+        else:
+            raise RuntimeError(f"Pinning memory is not supported for {place}")
+
     if in_dynamic_mode():
         shape = paddle.utils.convert_shape_to_list(shape)
-        return _C_ops.randint(
-            low, high, shape, dtype, _current_expected_place(), out=out
-        )
+        tensor = _C_ops.randint(low, high, shape, dtype, place, out=out)
+        if requires_grad is True:
```
Comment (zhwesky2010) [inline python/paddle/tensor/random.py]: pir下也需要处理requires_grad

**Comment by zhwesky2010** [comment]
infoflow 2026-03-02 21-00-38

pin_memory没覆盖到，需要测下pin_memory=True

File: python/paddle/tensor/random.py [hunk 4]
```diff
@@ -1765,6 +1765,35 @@ def uniform_(
...
+    low: int = 0,
+    high: int | None = None,
+    shape: ShapeLike = [1],
+    dtype: DTypeLike | None = None,
+    name: str | None = None,
+    *,
+    out: Tensor | None = None,
+    device: PlaceLike | None = None,
+    requires_grad: bool = False,
+    pin_memory: bool = False,
```
Comment (zhwesky2010) [inline python/paddle/tensor/random.py:1742]: 签名顺序写成 `device, pin_memory, requires_grad` 吧

其他的随机数也看下，顺便也改下

File: python/paddle/tensor/random.py [hunk 5]
```diff
@@ -1774,6 +1803,9 @@ def randint(
     name: str | None = None,
     *,
     out: Tensor | None = None,
+    device: PlaceLike | None = None,
+    requires_grad: bool = False,
+    pin_memory: bool = False,
```
Comment (zhwesky2010) [inline python/paddle/tensor/random.py:1771]: pin_memory位于device后面，只是一个代码风格问题。

其他的随机数API也检查下

File: test/legacy_test/test_randint_op.py [hunk 6]
```diff
@@ -360,5 +384,51 @@ def test_size_alias_old_static_mode(self):
             self.assertTrue(np.all(outs[0] >= 0) and np.all(outs[0] < 5))
 
 
+class TestRandintDeviceRequiresGradPinMemory(unittest.TestCase):
+    def test_device_cpu(self):
```
Comment (zhwesky2010) [inline test/legacy_test/test_randint_op.py:388]: out没测，可以再测全点，确保都覆盖上。下个PR顺带提交吧

Image

**Review by zhwesky2010 — DISMISSED** [review]
LGTM，小问题先合入

**Comment by zhwesky2010** [comment]
infoflow 2026-03-04 18-40-28

示例代码检查没通过

File: python/paddle/tensor/random.py [hunk 7]
```diff
@@ -1720,6 +1720,35 @@ def uniform_(
     return _C_ops.uniform_inplace_(x, min, max, seed, 0, 0, 1.0)
 
 
+@overload
```
Comment (zhwesky2010) [inline python/paddle/tensor/random.py:1732]: 还有一些示例代码的问题

Image

这个应该是size关键字没匹配上list(int)，再加个overload吧

---
Image

这个很奇怪，按道理ShapeLike匹配list(int)，DTypeLike匹配str。 @ShigureNyako 知道原因吗

---
Image

这个貌似没办法了，int只支持传入float/bool/complex，是不是得跳过了？@ShigureNyako

Comment (SigureMo) [reply python/paddle/tensor/random.py:1732]: 啊……你能看图片？

Comment (SigureMo) [reply python/paddle/tensor/random.py:1732]: 第二个：`DTypeLike` 不包含 `str`，只包含 `Literal["int32", "int8", ...]` 这种，因为这里可以提供更好的提示，在类型推导较好的工具（比如 `pyright` 或者 `ty`）这里是不会出问题的，mypy 现在不太行

这里建议 `type: ignore[call-overload]`，将来用 `ty` 就没这个问题了

第三个：这里的 `Union[float, bool, complex]` 是 `Tensor.item()`……关键是 `int(complex)` 是不合法的吧？这……

这里也建议 `type: ignore[arg-type]`

@ShigureNyako 笨哦，还不如我呢

**Review by zhwesky2010 — DISMISSED** [review]
LGTM

File: python/paddle/tensor/random.py [hunk 8]
```diff
@@ -1720,6 +1720,64 @@ def uniform_(
...
+
+
+@overload
+def randint(
+    low: int = 0,
+    high: int | None = None,
+    *,
+    size: ShapeLike,
+    dtype: DTypeLike | None = None,
+    name: str | None = None,
```
Comment (zhwesky2010) [inline python/paddle/tensor/random.py:1752]: 按上面的overload形式，直接把shape改成size吧

根据API的实际实现，name及之前的参数均为位置参数。

**Comment by SigureMo** [comment]
static-check 还没过呢，本地也可以自己跑一下 CI 里给的命令测测

**Review by zhwesky2010 — COMMENTED** [review]
Image

File: python/paddle/nn/functional/common.py [hunk 9]
```diff
@@ -2767,7 +2767,7 @@ class centers and the shape of sampled_class_center will be [num_positive_class_
         >>> # num_classes of each GPU can be different, e.g num_classes_list = [10, 8]
         >>> num_classes_list = [10, 10]
         >>> num_classes = paddle.sum(paddle.to_tensor(num_classes_list))
-        >>> label = paddle.randint(low=0, high=num_classes.item(), shape=[batch_size], dtype='int64')  # type: ignore[arg-type]
+        >>> label = paddle.randint(low=0, high=int(num_classes.item()), size=[batch_size], dtype='int64')  # type: ignore[call-overload]
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/common.py]: 从报错日志里，可以看出这里的报错码是 "arg-type"，所以你要写：`# type: ignore[arg-type]`

Image

File: python/paddle/nn/functional/loss.py [hunk 10]
```diff
@@ -2415,7 +2415,7 @@ def margin_cross_entropy(
             >>> num_class_per_card = [4, 8]
             >>> num_classes = paddle.sum(paddle.to_tensor(num_class_per_card))
 
-            >>> label = paddle.randint(low=0, high=num_classes.item(), shape=[batch_size], dtype='int64')  # type: ignore[arg-type]
+            >>> label = paddle.randint(low=0, high=int(num_classes.item()), size=[batch_size], dtype='int64')  # type: ignore[call-overload]
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py]: `# type: ignore[arg-type]`

File: test/legacy_test/test_api_compatibility.py [hunk 11]
```diff
@@ -2033,6 +2033,117 @@ def test_dygraph_axis(self):
...
 
+class TestRandintAPI_Compatibility(unittest.TestCase):
+    def test_dygraph_basic(self):
+        paddle.disable_static()
+        x = paddle.randint(high=10, shape=[2, 3])
+        self.assertEqual(x.shape, [2, 3])
+        self.assertTrue(x.stop_gradient)
+        paddle.enable_static()
+
+    def test_dygraph_size_alias(self):
```
Comment (zhwesky2010) [inline test/legacy_test/test_api_compatibility.py]: 测试case合并下吧，只用写两个case，每个case包含多种情况

**Review by zhwesky2010 — COMMENTED** [review]
这样改下应该可以过了

File: python/paddle/nn/functional/common.py (hunk 9 above)
Comment (SigureMo) [reply python/paddle/nn/functional/common.py]: 本地测一下再提交吧，如果是两个都有的话，这里需要写 `# type: ignore[call-overload, arg-type]`

**Review by zhwesky2010 — DISMISSED** [review]

File: python/paddle/distributed/auto_parallel/static/tuner/to_distributed_api_patterns.py [hunk 12]
```diff
@@ -312,7 +312,7 @@ def build(self):
         cos = paddle.randn(self.shape_configs.cos_shape)
         sin = paddle.randn(self.shape_configs.sin_shape)
         position_ids = paddle.randint(
-            low=1, shape=self.shape_configs.input_shape
+            low=1, size=self.shape_configs.input_shape
```
Comment (zhwesky2010) [inline python/paddle/distributed/auto_parallel/static/tuner/to_distributed_api_patterns.py:315]: 这些不是示例代码，应该不用改吧，不影响overload的添加

只用改示例代码

File: python/paddle/tensor/random.py [hunk 13]
```diff
@@ -1815,8 +1852,24 @@ def randint(
             [7])
             >>> # doctest: -SKIP
 
+            >>> # example 6:
+            >>> # Use 'size' as an alias for 'shape'
+            >>> out6 = paddle.randint(high=10, size=[2, 3])
+            >>> print(out6.shape)
+            [2, 3]
```
Comment (zhwesky2010) [inline python/paddle/tensor/random.py]: 你这个地方测试没有通过
Image

查看下log，搜下Failed 关键字

这个Static-check，目前看很难通过了，还是本地测试下提升效率。

Comment (zhwesky2010) [reply python/paddle/tensor/random.py]: @SigureMo 本地测试是跑Paddle/tools/type_checking.py脚本吗，这个有没自行类型检查的README

Comment (SigureMo) [reply python/paddle/tensor/random.py]: > static-check 还没过呢，本地也可以自己跑一下 CI 里给的命令测测

类型检查 CI 里挂了都会说怎么跑的……直接看 CI 报错就好了

现在类型检查已经过了，是示例代码检查挂了，这个就更简单了，自己把代码 copy 出去跑一遍看看输出就好了……这个不跑的话，输出不会是手填的吧？

Comment (zhwesky2010) [inline python/paddle/tensor/random.py]: 这个地方是`paddle.Size([2, 3])`

**Comment by zhwesky2010** [comment]
@Manfredss 修改一下CI问题

**Comment by zhwesky2010** [comment]
处理一下冲突，test_api_compatibility.py拆分了。并且按目前的测试模板规范来写吧：https://github.com/PaddlePaddle/docs/blob/develop/docs/dev_guides/coding_agent/api_compatibility/python-decorator.md#step-5-%E6%B7%BB%E5%8A%A0%E6%B5%8B%E8%AF%95%E7%94%A8%E4%BE%8B

File: test/legacy_test/test_api_compatibility_part3.py [hunk 14]
```diff
@@ -662,6 +662,72 @@ def test_static_Compatibility(self):
...
+                high=10, shape=[2, 3], requires_grad=False
+            )
+            self.assertTrue(x_no_grad.stop_gradient)
+            # size alias
+            x_size = paddle.randint(high=10, size=[2, 3])
+            self.assertEqual(x_size.shape, [2, 3])
+            # dtype string
+            x_dtype = paddle.randint(high=10, shape=[3], dtype='int32')
+            exe = paddle.static.Executor(paddle.CPUPlace())
+            exe.run(startup)
```
Comment (zhwesky2010) [inline test/legacy_test/test_api_compatibility_part3.py:726]: 不用run(startup)

File: test/legacy_test/test_api_compatibility_part3.py [hunk 15]
```diff
@@ -662,6 +662,72 @@ def test_static_Compatibility(self):
                 np.testing.assert_allclose(out, expected, rtol=1e-5)
 
 
+# Test randint compatibility
+class TestRandintAPI(unittest.TestCase):
```
Comment (zhwesky2010) [inline test/legacy_test/test_api_compatibility_part3.py:666]: 这个单测不是很规范，改成按以下来写：

https://github.com/PaddlePaddle/docs/blob/develop/docs/dev_guides/coding_agent/api_compatibility/python-decorator.md#step-5-%E6%B7%BB%E5%8A%A0%E6%B5%8B%E8%AF%95%E7%94%A8%E4%BE%8B 

不用自由发挥，按条目填上去。注意这里是 **全部关键字参数、全部位置参数、全部默认参数**

File: python/paddle/nn/layer/layers.py [hunk 16]
```diff
@@ -265,27 +253,22 @@ def _parse_to_args(*args, **kwargs):
...
-
-    # Resolve blocking
     if non_blocking is not None:
+        if not isinstance(non_blocking, bool):
+            raise ValueError("non_blocking value error, must be True or False")
         blocking = not non_blocking
-    elif blocking is None:
+    elif blocking is not None:
+        if not isinstance(blocking, bool):
+            raise ValueError(
```
Comment (ShigureNyako) [inline python/paddle/nn/layer/layers.py]: 这里把 `blocking=1` 的报错从原来的断言路径改成了 `ValueError`，已经把 `test_base_layer.TestLayerTo` 在多套 CI 上稳定打挂了：日志都停在 `test/legacy_test/test_base_layer.py:448` 的 `self.assertRaises(AssertionError, self.linear.to, blocking=1)`，例如 Linux-CPU / Mac-CPU / Windows-OPENBLAS / Linux-XPU / Linux-DCU 都是同一个栈。\n\n这部分也和当前 PR 的 `paddle.randint` 兼容性主题无关，建议先恢复 `to()` 现有行为，或者单独拆 PR 处理。

**Comment by Manfredss** [comment]
> 本轮先卡两个阻塞点：\n\n1. `python/paddle/nn/layer/layers.py` 里的 `to()` 行为改动把 `test_base_layer` 在多平台打挂了，我已在对应代码处留了 inline comment。建议不要把这部分混在 `paddle.randint` 的 PR 里。\n2. `Static-Check / Test` 也失败了。日志里 `tools/check_api_yaml_same.py` 后面列出了 `paddle.DataParallel.to`、`paddle.Tensor.to`、`paddle.rand`、`paddle.randint`、`paddle.randn`、`paddle.randperm` 等多处 API / document diff，说明当前提交已经超出本 PR 主题。\n\n建议先把变更范围收敛到 `randint` 必需部分，再继续看后续 review。

明白。之前周会时让我顺便加在这个 PR 里一起提交，因为对应的 PR 已经合入。我重新提一个 PR

**Review by zhwesky2010 — COMMENTED** [review]
小问题下个PR改下吧

**Review by zhwesky2010 — APPROVED** [review]

**Comment by zhwesky2010** [comment]
LGTM

**Review by SigureMo — APPROVED** [review]
